runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -184,8 +187,73 @@
     void add(Iterable<WindowedValue<ElemT>> values);
   }
 
+  /** The set of {@link PTransform PTransforms} that execute a UDF. Useful for some enforcements. */
+  private static final Set<Class<? extends PTransform>> CONTAINS_UDF =
+      ImmutableSet.of(
+          Read.Bounded.class, Read.Unbounded.class, ParDo.Bound.class, ParDo.BoundMulti.class);
+
+  enum Enforcement {
+    ENCODABILITY {
+      @Override
+      public boolean appliesTo(PTransform<?, ?> transform) {
+        return true;
+      }
+    },
+    IMMUTABILITY {
+      @Override
+      public boolean appliesTo(PTransform<?, ?> transform) {
+        return CONTAINS_UDF.contains(transform.getClass());
+      }
+    };
+
+    public abstract boolean appliesTo(PTransform<?, ?> transform);
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Utilities for creating enforcements
+    public static Set<Enforcement> enabled(DirectOptions options) {
+      EnumSet<Enforcement> enabled = EnumSet.noneOf(Enforcement.class);
+      if (options.isEnforceEncodability()) {
+        enabled.add(ENCODABILITY);
+      }
+      if (options.isEnforceImmutability()) {
+        enabled.add(IMMUTABILITY);
+      }
+      return Collections.unmodifiableSet(enabled);
+    }
+
+    public static BundleFactory bundleFactoryFor(Set<Enforcement> enforcements) {
+      BundleFactory bundleFactory =
+          enforcements.contains(Enforcement.ENCODABILITY)
+              ? CloningBundleFactory.create()
+              : ImmutableListBundleFactory.create();
+      if (enforcements.contains(Enforcement.ENCODABILITY)) {
+        bundleFactory = ImmutabilityCheckingBundleFactory.create(bundleFactory);
```

Comment:
This is a specific requirement to the `CloningBundleFactory` - because the underlying bundle gets the clones of elements, it must not wrap the `ImmutabilityCheckingBundleFactory` - otherwise the immutability-checked element is not the bundle element.